src/lua/datetime.lua

```diff
@@ -0,0 +1,611 @@
+    end
+
+    local ts = obj.timestamp
+    if ts ~= nil then
+        local sec_int, fraction
+        sec_int, fraction = math_modf(ts)
+        -- if there is one of nsec, usec, msec provided
+        -- then ignore fraction in timestamp
+        -- otherwise - use nsec, usec, or msec
+        if count_usec == 0 then
+            nsec = fraction * 1e9
+        else
+            error('only integer values allowed in timestamp '..
+                  'if nsec, usec, or msecs provided', 2)
+        end
+
+        self.epoch = sec_int
+        self.nsec = nsec
+
+        return self
+    end
+
+    local offset0 = self.tzoffset
+    local offset = obj.tzoffset
+    if offset ~= nil then
+        offset = get_timezone(offset)
+        check_range(offset, -720, 720, 'tzoffset')
+    end
+
+    if obj.tz ~= nil then
+        nyi('tz')
+    end
+
+    -- .year, .month, .day
+    if ymd then
+        datetime_ymd_update(self, y or y0, M or M0, d or d0, offset or offset0)
+    end
+
+    -- .hour, .minute, .second
+    if hms then
+        datetime_hms_update(self, h or h0, m or m0, sec or sec0, offset or offset0)
+    end
+
+    self.epoch, self.nsec = normalize_nsec(self.epoch, self.nsec)
+
+    if offset ~= nil then
+        self.tzoffset = offset
+    end
+
+    return self
+end
+
+local function datetime_strftime(self, fmt)
+    local strfmt_sz = 128
+    local buff = ffi.new('char[?]', strfmt_sz)
+    check_date(self, "datetime.strftime()")
+    check_str(fmt, "datetime.strftime()")
+    builtin.datetime_strftime(self, buff, strfmt_sz, fmt)
+    return ffi.string(buff)
+end
+
+local function datetime_format(self, fmt)
+    if fmt ~= nil then
+        return datetime_strftime(self, fmt)
+    else
+        return datetime_tostring(self)
+    end
+end
+
+ffi.metatype(datetime_t, {
+    __tostring = datetime_tostring,
+    __serialize = datetime_serialize,
+    __eq = datetime_eq,
+    __lt = datetime_lt,
+    __le = datetime_le,
+    __index = {
+        epoch = function(self) return self.epoch end,
+        timestamp = function(self) return self.epoch + self.nsec / 1e9 end,
+
+        usec = function(self) return self.nsec / 1e3 end,
+        msec = function(self) return self.nsec / 1e6 end,
+
+        dt = function(self) return local_dt(self) end,
+        year = function(self) return builtin.tnt_dt_year(local_dt(self)) end,
+        month = function(self) return builtin.tnt_dt_month(local_dt(self)) end,
+
+        yday = function(self) return builtin.tnt_dt_doy(local_dt(self)) end,
+        wday = function(self)
+            return ffi.cast('int32_t', builtin.tnt_dt_dow(local_dt(self))) + 1
+        end,
+        day = function(self)
+            return builtin.tnt_dt_dom(local_dt(self))
+        end,
+        hour = function(self) return math_floor((local_secs(self) / 3600) % 24) end,
+        min = function(self) return math_floor((local_secs(self) / 60) % 60) end,
+        sec = function(self) return self.epoch % 60 end,
+
```

Comment:
This looks broken:
```
tarantool> d.sec
---
- 'function: 0x0103b06ee0'
...

tarantool> d.nsec
---
- 622745000
...
```
Some members are functions, some are values. I don't know what is the way Mons and unera want, but certainly it must be something used for all members, not for some.

Reply:
`nsec` as value is not from `__index`, where it was wrongly created as function, but rather from ffi definition. There is no way to create function using the same name as ffi structure field. I've removed `nsec` as function - there is no known way to make it work as function (if only using different name than field)

Reply:
In rfc all accessors considered as "fields", not as methods

Reply:
Ok, here is modified version:

```diff
diff --git a/src/lua/datetime.lua b/src/lua/datetime.lua
index 6cd66fbe7..a9758e66a 100644
--- a/src/lua/datetime.lua
+++ b/src/lua/datetime.lua
@@ -568,13 +568,7 @@ local function datetime_format(self, fmt)
     end
 end
 
-ffi.metatype(datetime_t, {
-    __tostring = datetime_tostring,
-    __serialize = datetime_serialize,
-    __eq = datetime_eq,
-    __lt = datetime_lt,
-    __le = datetime_le,
-    __index = {
+local datetime_index_fields = {
         timestamp = function(self) return self.epoch + self.nsec / 1e9 end,
 
         year = function(self) return builtin.tnt_dt_year(local_dt(self)) end,
@@ -591,11 +585,27 @@ ffi.metatype(datetime_t, {
         sec = function(self) return self.epoch % 60 end,
         usec = function(self) return self.nsec / 1e3 end,
         msec = function(self) return self.nsec / 1e6 end,
+}
 
+local datetime_index_functions = {
         format = datetime_format,
         totable = datetime_totable,
         set = datetime_set,
-    }
+}
+
+local function datetime_index(self, key)
+    local handler_field = datetime_index_fields[key]
+    return handler_field ~= nil and handler_field(self) or
+           datetime_index_functions[key]
+end
+
+ffi.metatype(datetime_t, {
+    __tostring = datetime_tostring,
+    __serialize = datetime_serialize,
+    __eq = datetime_eq,
+    __lt = datetime_lt,
+    __le = datetime_le,
+    __index = datetime_index,
 })
 
 return setmetatable(
diff --git a/test/app-tap/datetime.test.lua b/test/app-tap/datetime.test.lua
index 6d0667d0e..a54584e46 100755
--- a/test/app-tap/datetime.test.lua
+++ b/test/app-tap/datetime.test.lua
@@ -235,24 +235,25 @@ test:test("Datetime string formatting detailed", function(test)
 end)
 
 test:test("__index functions()", function(test)
-    test:plan(11)
+    test:plan(12)
     -- 2000-01-29T03:30:12Z'
     local T = date.new{sec = 12, min = 30, hour = 3,
                        tzoffset = 0,  day = 29, month = 1, year = 2000,
                        nsec = 123000000}
 
-    test:is(T:min(), 30, 'T:min()')
-    test:is(T:wday(), 7, 'T:wday()')
-    test:is(T:yday(), 29, 'T:yday()')
-    test:is(T:year(), 2000, 'T:year()')
-    test:is(T:month(), 1, 'T:month()')
-    test:is(T:day(), 29, 'T:day()')
-    test:is(T:hour(), 3, 'T:hour()')
-    test:is(T:min(), 30, 'T:min()')
-    test:is(T:sec(), 12, 'T:sec()')
-
-    test:is(T:usec(), 123000, 'T:usec()')
-    test:is(T:msec(), 123, 'T:msec()')
+    test:is(T.min, 30, 'T.min')
+    test:is(T.wday, 7, 'T.wday')
+    test:is(T.yday, 29, 'T.yday')
+    test:is(T.year, 2000, 'T.year')
+    test:is(T.month, 1, 'T.month')
+    test:is(T.day, 29, 'T.day')
+    test:is(T.hour, 3, 'T.hour')
+    test:is(T.min, 30, 'T.min')
+    test:is(T.sec, 12, 'T.sec')
+
+    test:is(T.nsec, 123000000, 'T.nsec')
+    test:is(T.usec, 123000, 'T.usec')
+    test:is(T.msec, 123, 'T.msec')
 end)
 
 test:test("totable{}", function(test)

```

Reply:
>  there is no known way to make it work as function (if only using different name than field)

But you just named the solution - in Lua structs you can use any names, they don't need to match the C struct's member names. Anyway, irrelevant here already.